File: proj-java/src/main/java/engine/Posts.java
 package engine;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Posts {
    private long ID;
    private String title;
    private long ownerID;
    private Owner autor;
    private short typeID;
    private long parentID;
    private LocalDate creation_date;
    private List<String> tag;
    private int score;
    private int comentCount;
    private int answerCount;

    /**
    *@brief Construtor por omissao de Posts
    */
    public Posts(){
        this.ID = 0;
        this.title = "";
        this.ownerID = 0;
        this.autor = new Owner();
        this.typeID = 0;
        this.parentID = 0;
        this.creation_date = LocalDate.now();
        this.tag = new ArrayList<>();
        this.score = 0;
        this.comentCount = 0;
        this.answerCount = 0;
    }
    /**
    *@brief Construtor por copia de Posts
    */
    public Posts(Posts p){
        this.ID = p.getID();
        this.title = p.getTitle();
        this.ownerID = p.getOwnerID();
        this.autor = p.getAutor();
        this.typeID = p.getTypeID();
        this.parentID = p.getParentID();
        this.creation_date = p.getCreation_date();
        this.tag = p.getTag();
        this.score = p.getScore();
        this.comentCount = p.getComentCount();
        this.answerCount = p.getAnswerCount();
    }
    /**
    *@brief Construtor parametrizado de Posts
    */
    public Posts(long id, String titulo, long ownerid, Owner author, short typeid, long parentid, LocalDate creationdate, List<String> tags, int score, int comentCount, int answerCount){
        this.ID = id;
        this.title = titulo;
        this.ownerID = ownerid;
        this.autor = author;
        this.typeID = typeid;
        this.parentID = parentid;
        this.creation_date = creationdate;
        this.tag = tags;
        this.score = score;
        this.comentCount = comentCount;
        this.answerCount = answerCount;
    }
    /**
    *Devolve o id do Post
    *
    * @return  o id do Post
    */
    public long getID() {
        return this.ID;
    }
    /**
         * Atualiza o id do Post
         * @param long
         */
    public void setID(long ID) {
        this.ID = ID;
    }
    /**
    *Devolve o titulo do Post
    *
    * @return  o titulo do Post
    */
    public String getTitle() {
        return this.title;
    }
    /**
         * Atualiza o titulo do Post
         * @param String
         */
    public void setTitle(String title) {
        this.title = title;
    }
    /**
    *Devolve o id do owner do post
    *
    * @return  o id do owner do post
    */
    public long getOwnerID() {
        return this.ownerID;
    }
    /**
         * Atualiza o id do owner do Post
         * @param long
         */
    public void setOwnerID(long ownerID) {
        this.ownerID = ownerID;
    }
    /**
    *Devolve o owner do post
    *
    * @return   o owner do post
    */
    public Owner getAutor() {
        return this.autor;
    }
    /**
         * Atualiza o owner do post
         * @param Owner
         */
    public void setAutor(Owner autor) {
        this.autor = autor.clone();
    }
    /**
    *Devolve o tipo do post
    *
    * @return  o tipo  do post
    */
    public short getTypeID() {
        return this.typeID;
    }
    /**
         * Atualiza o tipo  do Post
         * @param short
         */
    public void setTypeID(short typeID) {
        this.typeID = typeID;
    }
    /**
    *Devolve o Parentid do post
    *
    * @return  o Parentid do post
    */
    public long getParentID() {
        return this.parentID;
    }
    /**
         * Atualiza o parentid do Post
         * @param long
         */
    public void setParentID(long parentID) {
        this.parentID = parentID;
    }
    /**
    *Devolve a data do post
    *
    * @return LocalDate
    */
    public LocalDate getCreation_date() {
        return this.creation_date;
    }
    /**
         * Atualiza a data  do Post
         * @param LocalDate
         */
    public void setCreation_date(LocalDate creation_date) {
        this.creation_date = creation_date;
    }
    /**
    *Devolve as tags do post
    *
    * @return List<String>
    */
    public List<String> getTag() {
      List<String> aux= new ArrayList<>();
      for(String s :this.tag){
          aux.add(s);
      }
        return aux;
    }
    /**
         * Atualiza as tags do Post
         * @param ArrayList<String>
         */
    public void setTag(ArrayList<String> tag) {
       this.tag= new ArrayList<>();
      for(String s :tag){
          this.tag.add(s);
      }
    }
    /**
    *Devolve o score do post
    *
    * @return int
    */
    public int getScore() {
        return this.score;
    }
    /**
         * Atualiza o Score do Post
         * @param int
         */
    public void setScore(int score) {
        this.score = score;
    }
    /**
    *Devolve o numero de comentarios do post
    *
    * @return int
    */
    public int getComentCount() {
        return this.comentCount;
    }
    /**
         * Atualiza o CommentCount do Post
         * @param int
         */
    public void setComentCount(int comentCount) {
        this.comentCount = comentCount;
    }
    /**
    *Devolve o numero de respostas do post
    *
    * @return int
    */
    public int getAnswerCount() {
        return this.answerCount;
    }
    /**
         * Atualiza o answerCount do Post
         * @param int
         */
    public void setAnswerCount(int answerCount) {
        this.answerCount = answerCount;
    }

    /**
    *Metodo clone
    *
    * @return Posts
    */
    public Posts clone(){
        return new Posts(this);
    }
    /**
    *Metodo equals
    *
    * @return bolean
    */
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || obj.getClass() != this.getClass()) return false;
        Posts p = (Posts) obj;
        return p.getID() == this.ID
                && p.getTitle().equals(this.title)
                && p.getOwnerID() == this.ownerID
                && p.getAutor().equals(this.autor)
                && p.getTypeID() ==this.typeID
                && p.getParentID() == this.parentID
                && p.getCreation_date().equals(this.creation_date)
                && p.getTag().equals(this.tag)
                && p.getScore() == this.score
                && p.getComentCount() == this.comentCount
                && p.getAnswerCount() == this.answerCount;
    }
    /**
    *Metodo toString
    *
    * @return String
    */
    public String toString(){
        StringBuilder sb = new StringBuilder();
        sb.append("\n").append("ID:" + this.ID);
        sb.append("\n").append("Title:" +this.title);
        sb.append("\n").append("OwnerID:" +this.ownerID);
        sb.append("\n").append("Autor:" +this.autor);
        sb.append("\n").append("TypeID" +this.typeID);
        sb.append("\n").append("ParentID:" +this.parentID);
        sb.append("\n").append("Creation Date:" +this.creation_date);
        sb.append("\n").append("Tag:" +this.tag);
        sb.append("\n").append("Score:" +this.score);
        sb.append("\n").append("Comment Count:" +this.comentCount);
        sb.append("\n").append("Answer:" +this.answerCount);
        return sb.toString();
    }
}

File: proj-c/src/posts.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "posts.h"
#include "assert.h"
#include "data.h"
#include "common.h"
#include "tag_ligada.h"

/**
* @file posts.c
* @author Grupo56
* @date 5 May 2018
* @brief Struct que guarda um post e toda a sua informacao
*/

	struct post {
		long id;
		char * title;
    long ownerId;
		Owner autor;
		short  typeId;
		long parentId;
		DATA creation_date;
		Tags tag;
		int score;
		int comentCount;
		int answerCount;
 };




/**
* @brief
* @param id id do post
* @param title titulo do post
* @param ownerId id do criador do post
* @param o informações do Owner
* @param typeId se o post é pergunta ou resposta
* @param parentId id do pai do post
* @param d Data
* @param t tags do post
* @param score score do post
* @param comentCount numero de comentarios
* @param answerCount numero de respostas
* @return
*/
Post createPost(long id, char * title,long ownerId, Owner o,short typeId,long parentId, DATA d,Tags t,int score, int comentCount,int answerCount) {

    Post p = malloc(sizeof(struct post));
    p->id =id;
    p->title = title;
    p->ownerId =ownerId;
    p->autor = o;
    p->typeId=typeId;
    p->parentId=parentId;
    p->creation_date=d;
    p->tag =t;
    p->score=score;
    p->comentCount=comentCount;
    p->answerCount=answerCount;
    return p;
}



/**
* @brief Procura o id do Post
* @param p Post
* @return Retorna o id do Post
*/
long get_id(Post p) {
    return p->id;
}


/**
* @brief Procura o titulo do Post
* @param p Post
* @return Retorna o titulo do Post
*/
char * get_title(Post p) {
    return mystrdup (p->title); //change to enum? (Acho que sim. Desta forma já garantimos a limitação necessária)
}


/**
* @brief Procura o id do Owner do Post
* @param p Post
* @return Retorna o ownerId
*/
long get_ownerId(Post p){
    return p->ownerId;
}

/**
* @brief Procura o nome do autor do Post
* @param p Post
* @return Retorna o autor do texto
*/
Owner get_autor(Post p) {
    return p->autor;
}

/**
* @brief Procura o tipo de Post que é,se é pergunta ou resposta
* @param p Post
* @return Retorna o typeId do Post
*/
short get_typeId(Post p) {
    return p->typeId;
}

/**
* @brief Procura o id do pai do Post
* @param p Post
* @return Retorna o parentId
*/
long get_parentId(Post p) {
    return p->parentId;
}

/**
* @brief Procura a data de criazao do post
* @param p Post
* @return Retorna a creationDate
*/
DATA get_CreationDate(Post p) {
    if (!p)return NULL;
    return clone_DATA (p->creation_date);
}

/**
* @brief Procura a lista de tags do Post, clonando para nao se perder informação
* @param p Post
* @return Retorna a lista de tags do post
*/
Tags get_Tags(Post p) {
    return clone_Tags(p->tag);
}

/**
* @brief Procura o score do post
* @param p Post
* @return Retorna o score do Post
*/
int get_score(Post p) {
    return p->score;
}

/**
* @brief Procura o inteiro relativo ao numero de comentarios do post
* @param p Post
* @return Retorna o comentCount do Post
*/
int get_comentCount(Post p){
	return p->comentCount;
}

/**
* @brief Procura o numero de perguntas do post
* @param p Post
* @return Retorna o answerCount do Post
*/
int get_answerCount(Post p){
	return p->answerCount;
}

/**
* @brief Vai ao Owner e retira a string relativa ao nome do autor
* @param p Post
* @return Retorna o nome dentro da struct Owner
*/
char* get_NAME_OwnerPost(Post p){
    return get_Oname(p->autor);
}

/**
* @brief
* @param
* @return 
*/
long get_ID_ownerPost(Post p){
        return get_Oid(p->autor);
}

int get_REPUTACION_ownerPost(Post p){
    return get_Oreputation(p->autor);
}
int get_DATE_DAY_post(Post p){
    return get_data_dia(p->creation_date);
}

int get_DATE_MONTH_post(Post p){
    return get_data_mes(p->creation_date);
}

int get_DATE_YEAR_post(Post p){
    return get_data_ano(p->creation_date);
}



void set_Owner(Post p,Owner o ){
    p->autor=o;
}

long getquestionID(Post p){
    if(p->typeId == 1){
        return p->id;
    }else if (p->typeId == 2){
        return p->parentId;
    }else
    return 0;
}



void free_post(Post p) {
    free(p->title);
    free_owner(p->autor);
    free_tag(p->tag);
    free_data(p->creation_date);
    free(p);
}
